File: top_interview_questions_easy/symmetric_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };       
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
      if(root == nullptr) return true;

      if(root -> left == nullptr && root -> right == nullptr) return true;
      if(root -> left == nullptr || root -> right == nullptr) return false;

      stack<TreeNode*> dfs;
      dfs.push(root);
      dfs.push(root);
      while(dfs.empty() == false) {
        TreeNode* t1 = dfs.top();
        dfs.pop();
        TreeNode* t2 = dfs.top();
        dfs.pop();

        if(t1 -> val != t2 -> val) {
          return false;
        }
        if(t1 -> left != nullptr && t2 -> right != nullptr) {
          dfs.push(t1 -> left);
          dfs.push(t2 -> right);
        }else if(t1 -> left != nullptr || t2 -> right != nullptr) {
          return false;
        }

        if(t1 -> right != nullptr && t2 -> left != nullptr) {
          dfs.push(t1 -> right);
          dfs.push(t2 -> left);
        }else if(t1 -> right != nullptr || t2 -> left != nullptr) {
          return false;
        }
      }
      return true;
    }
};

File: top_interview_questions_easy/symmetric_tree.py
"""
 root를 확인하고 왼쪽 서브트리와 오른쪽 서브트리를 확인한다.
 대칭인 점을 확인하려면 왼쪽 서브트리와 오른쪽 서브트리를 서로 반대로 비교해야한다.
 두 서브트리의 root를 확인한 후 왼쪽 서브트리의 left가 오른쪽 서브트리의 right와 같은지.
 왼쪽 서브트리의 right가 오른쪽 서브트리의 left와 같은지 비교한다.
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def isSubTreeSymmetric(left, right):
            if not left and not right:
                return True
            elif not left or not right:
                return False
            else:
                return left.val == right.val and isSubTreeSymmetric(left.left, right.right) and isSubTreeSymmetric(left.right, right.left)
        return isSubTreeSymmetric(root, root)